File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
 int d=0;
 int max(int l, int r)
 {
    if(l>r){
        return l;
    }
    else{
        return r;
    }
 }
 int height(struct TreeNode* root){
    if(root==NULL){
        return 0;
    }
    int left=height(root->left);
    int right=height(root->right);
    d=max(d,left+right);
    return 1+max(left,right);
 }
int diameterOfBinaryTree(struct TreeNode* root) {
 d=0;
height(root);
return d;
    
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        d=0
        def height(node):
            nonlocal d  #we declared d=0 outiside the function so it avoids it becoming a new variable
            if node is None:
                return 0
            left=height(node.left)
            right=height(node.right)
            d=max(d,left+right)
            h=max(left,right)+1
            return h
        height(root)
        return d
        